qutip/nonmarkov/dlheom_solver.py

```diff
@@ -0,0 +1,595 @@
+                            ss_conf)
+
+        if stats: start_integ_conf = timeit.default_timer()
+
+        r = scipy.integrate.ode(cy_ode_rhs)
+
+        r.set_f_params(L_helems.data, L_helems.indices, L_helems.indptr)
+        r.set_integrator('zvode', method=options.method, order=options.order,
+                         atol=options.atol, rtol=options.rtol,
+                         nsteps=options.nsteps, first_step=options.first_step,
+                         min_step=options.min_step, max_step=options.max_step)
+
+        if stats:
+            time_now = timeit.default_timer()
+            stats.add_timing('Liouvillian contruct',
+                             time_now - start_integ_conf,
+                            ss_conf)
+            if ss_conf.total_time is None:
+                ss_conf.total_time = time_now - start_config
+            else:
+                ss_conf.total_time += time_now - start_config
+
+        self._ode = r
+        self._N_he = N_he
+        self._sup_dim = sup_dim
+        self._configured = True
+
+    def run(self, rho0, tlist):
+        """
+        Function to solve for an open quantum system using the
+        HEOM model.
+
+        Parameters
+        ----------
+        rho0 : Qobj
+            Initial state (density matrix) of the system.
+
+        tlist : list
+            Time over which system evolves.
+
+        Returns
+        -------
+        results : :class:`qutip.solver.Result`
+            Object storing all results from the simulation.
+        """
+
+        start_run = timeit.default_timer()
+
+        sup_dim = self._sup_dim
+        stats = self.stats
+        r = self._ode
+
+        if not self._configured:
+            raise RuntimeError("Solver must be configured before it is run")
+        if stats:
+            ss_conf = stats.sections.get('config')
+            if ss_conf is None:
+                raise RuntimeError("No config section for solver stats")
+            ss_run = stats.sections.get('run')
+            if ss_run is None:
+                ss_run = stats.add_section('run')
+
+        # Set up terms of the matsubara and tanimura boundaries
+        output = Result()
+        output.solver = "hsolve"
+        output.times = tlist
+        output.states = []
+
+        if stats: start_init = timeit.default_timer()
+        output.states.append(Qobj(rho0))
+        rho0_flat = rho0.full().ravel('F') # Using 'F' effectively transposes
+        rho0_he = np.zeros([sup_dim*self._N_he], dtype=complex)
+        rho0_he[:sup_dim] = rho0_flat
+        r.set_initial_value(rho0_he, tlist[0])
+
+        if stats:
+            stats.add_timing('initialize',
+                             timeit.default_timer() - start_init, ss_run)
+            start_integ = timeit.default_timer()
+
+        dt = np.diff(tlist)
+        n_tsteps = len(tlist)
+        for t_idx, t in enumerate(tlist):
+            if t_idx < n_tsteps - 1:
+                r.integrate(r.t + dt[t_idx])
```

Comment:
After integrating, you should check `r.successful()`. If the integrator fails, it does not raise an error but an `IntegrationWarning` and the state `r.y` is wrong, but often it still seems plausible.

The main error that you can expect is that too much work is done in one step, it this case it will stop the integration where it is when it reach the maximum work. `r.t` will be less than what you expect and the state will be at that time. Since you don't check, you may be saving the `state`/`expect` with a small time shift than what you expect.

Reply:
Thank you! These are important checks to have. Checks and tests for them added.